subscription_manager.go

```diff
@@ -0,0 +1,271 @@
+package river
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/riverqueue/river/internal/baseservice"
+	"github.com/riverqueue/river/internal/jobcompleter"
+	"github.com/riverqueue/river/internal/jobstats"
+	"github.com/riverqueue/river/internal/util/sliceutil"
+	"github.com/riverqueue/river/rivertype"
+)
+
+type subscriptionManager struct {
+	baseservice.BaseService
+
+	subscribeCh <-chan []jobcompleter.CompleterJobUpdated
+	wg          sync.WaitGroup
+
+	statsMu        sync.Mutex // protects stats fields
+	statsAggregate jobstats.JobStatistics
+	statsNumJobs   int
+
+	mu               sync.Mutex // protects subscription fields
+	subscriptions    map[int]*eventSubscription
+	subscriptionsSeq int // used for generating simple IDs
+}
+
+func newSubscriptionManager(archetype *baseservice.Archetype, subscribeCh <-chan []jobcompleter.CompleterJobUpdated) *subscriptionManager {
+	return baseservice.Init(archetype, &subscriptionManager{
+		subscribeCh:   subscribeCh,
+		subscriptions: make(map[int]*eventSubscription),
+	})
+}
+
+// ResetSubscribeChan is used to change the channel that the subscription
+// manager listens on. It must only be called when the subscription manager is
+// stopped.
+func (sm *subscriptionManager) ResetSubscribeChan(subscribeCh <-chan []jobcompleter.CompleterJobUpdated) {
+	sm.subscribeCh = subscribeCh
+}
+
+func (sm *subscriptionManager) Start(ctx context.Context) error {
+	sm.wg.Add(1)
```

Comment:
Can we make this use the normal stop convention instead with `StartInit`? Protects against double starts, but also gives the client a way to check whether all its services are started.

Reply:
I think I've made the appropriate changes for this to use `startstop.BaseStartStop`, please take a look to see if I did so correctly!